File: src/Core/Models/AccountInfo.cs
namespace CyberPatriotAutomation.Core.Models;

/// <summary>
/// Represents user account information and permissions
/// </summary>
public class AccountInfo
{
    public string Username { get; set; } = string.Empty;
    public string FullName { get; set; } = string.Empty;
    public bool IsEnabled { get; set; }
    public bool IsAdmin { get; set; }
    public bool PasswordRequired { get; set; }
    public bool PasswordNeverExpires { get; set; }
    public bool CannotChangePassword { get; set; }
    public DateTime? LastLogon { get; set; }
    public DateTime? PasswordLastSet { get; set; }
    public List<string> GroupMemberships { get; set; } = new();
}

/// <summary>
/// Security standards for account permissions
/// </summary>
public static class AccountSecurityStandards
{
    /// <summary>
    /// Guest account should be disabled
    /// </summary>
    public const bool GuestAccountShouldBeDisabled = true;

    /// <summary>
    /// Administrator account should be renamed from default
    /// </summary>
    public const bool AdminAccountShouldBeRenamed = true;

    /// <summary>
    /// All users should have password required
    /// </summary>
    public const bool PasswordShouldBeRequired = true;

    /// <summary>
    /// Passwords should expire (not set to never expire)
    /// </summary>
    public const bool PasswordShouldExpire = true;

    /// <summary>
    /// Maximum days of inactivity before account should be reviewed
    /// </summary>
    public const int MaxInactiveDays = 90;

    /// <summary>
    /// Known default/insecure usernames to check
    /// </summary>
    public static readonly string[] InsecureUsernames = new[]
    {
        "Guest",
        "DefaultAccount",
        "WDAGUtilityAccount",
    };

    /// <summary>
    /// Built-in admin account name to check for rename
    /// </summary>
    public const string DefaultAdminName = "Administrator";
}

File: rust/src/models/account_info.rs
use chrono::{DateTime, Local};

/// Represents user account information and permissions.
#[derive(Debug, Default, Clone)]
pub struct AccountInfo {
    pub username: String,
    pub full_name: String,
    pub is_enabled: bool,
    pub is_admin: bool,
    pub password_required: bool,
    pub password_never_expires: bool,
    pub cannot_change_password: bool,
    pub last_logon: Option<DateTime<Local>>,
    pub password_last_set: Option<DateTime<Local>>,
    pub group_memberships: Vec<String>,
}

/// Security standards for account permissions.
pub struct AccountSecurityStandards;

impl AccountSecurityStandards {
    /// Guest account should be disabled.
    pub const GUEST_ACCOUNT_SHOULD_BE_DISABLED: bool = true;
    /// Administrator account should be renamed from default.
    pub const ADMIN_ACCOUNT_SHOULD_BE_RENAMED: bool = true;
    /// All users should have password required.
    pub const PASSWORD_SHOULD_BE_REQUIRED: bool = true;
    /// Passwords should expire (not set to never expire).
    pub const PASSWORD_SHOULD_EXPIRE: bool = true;
    /// Maximum days of inactivity before account should be reviewed.
    pub const MAX_INACTIVE_DAYS: i64 = 90;

    /// Known default/insecure usernames to check.
    pub const INSECURE_USERNAMES: &'static [&'static str] =
        &["Guest", "DefaultAccount", "WDAGUtilityAccount"];

    /// Built-in admin account name to check for rename.
    pub const DEFAULT_ADMIN_NAME: &'static str = "Administrator";
}
